### ISSUE TITLE
upsertGraph (with relate) on HasMany example?

### ISSUE THREAD

**Issue by booboothefool** [Vincit/objection.js#1259]
So previous I was using surrogate keys ManyToMany where a user can speak many languages:

profile -> profile_languages -> language

```
profile
---
id
1

profile_languages
---
id  profile_id  language_id
1   1           1
2   1           2

language
---
id type
1  English
2  Chinese
```

```
// Profile model
languages: { relation: Model.ManyToManyRelation, modelClass: language, join: { from: 'profile.id', through: { from: 'profile_languages.profile_id', to: 'profile_languages.language_id' }, to: 'language.id' } },
```

```
relate: ['languages'],
unrelate: ['languages']
```

And I could upsert like:
```
profile: {
  id: 1,
  languages: [{ id: 1 }, { id: 2 }],
},
```

---

Now I am trying to get rid of the extra join, and having trouble upserting.

profile -> profile_languages

```
profile
---
id
1

profile_languages
---
id  profile_id  language_id
1   1           English
2   1           Chinese

(now no longer need to join to language because we already have the name from natural key)

language
---
id
English
Chinese
```

(I have also tried with and without a composite key):
```
profile_languages
---
profile_id  language_id
1           English
1           Chinese

profile_languages
---
id  profile_id  language_id
1   1           English
2   1           Chinese
```

```
class profile_languages extends Model {
    static get tableName() { return 'profile_languages' }

    // if using composite key
    static get idColumn() {
        return ['profile_id', 'language_id'];
    }
}
```

```
// Profile model
languages: { relation: Model.HasManyRelation, modelClass: profile_languages, join: { from: 'profile.id', to: 'profile_languages.profile_id' } },
```

I tried defining:
```
relate: ['languages'],
unrelate: ['languages']
```

I would like to upsert relate something like this:
```
profile: {
  id: 1,
  languages: [{ language_id: 'English' }, { language_id: 'Chinese' }],
},
```

In summary, I am trying to go from (surrogate keys) ManyToMany -> (natural keys) HasMany, and still be able to upsertGraph on the HasMany. I think I just need to see an example of the upsertGraph with HasMany.

**Comment by koskimas** [comment]
`relate` and `unrelate` make no sense in this case. What `relate` does for `HasManyRelation` is that it updates the related tables foreign key, so in this case both the `language_id` and `profile_id`, but you don't have `language_id` in `profile` so that's going to be pretty much broken.

First of all, **consider not using upsertGraph**! I mean there are use cases for it, but don't use it as a silver bullet for using relational database like a document database. This particular problem would be pretty trivial without `upsertGraph`.

Now to the problem: I think this should work if you just remove `relate` and `unrelate` options. What you really want with `HasManyRelation`, where `Profile` has the ownership of the items, is to delete and insert rows. So what happens if you just remove those options?

**Comment by koskimas** [comment]
And also remove the composite key. You only need `id` to be the indentifier here.

**Comment by booboothefool** [comment]
For some reason this HasMany was not as intuitive to me as getting the ManyToMany to work, but it does work with your recommendations! Thank you so much, here is the final solution:

```
profile
---
id
1

profile_languages
---
id  profile_id(fk)  language_id(fk)
1   1               English
2   1               Chinese

language
---
id
English
Chinese
```

```
class profile_languages extends Model {
    static get tableName() { return 'profile_languages' }

    // default idColumn = id :) 
}
```

```
// Profile model
languages: { relation: Model.HasManyRelation, modelClass: profile_languages, join: { from: 'profile.id', to: 'profile_languages.profile_id' } },
```

```
didn't use relate :)
// relate: ['languages'],
// unrelate: ['languages']
```

```
// upsertGraph
profile: {
  id: 1,
  languages: [{ language_id: 'English' }, { language_id: 'Chinese' }],
},
```

**Comment by koskimas** [comment]
👍
